pinot-core/src/main/java/org/apache/pinot/core/transport/InstanceRequestHandler.java

```diff
@@ -60,67 +64,107 @@ public InstanceRequestHandler(QueryScheduler queryScheduler, ServerMetrics serve
-          ctx.writeAndFlush(Unpooled.wrappedBuffer(responseBytes)).addListener(f -> {
-            long sendResponseEndTimeMs = System.currentTimeMillis();
-            int sendResponseLatencyMs = (int) (sendResponseEndTimeMs - sendResponseStartTimeMs);
-            _serverMetrics.addMeteredGlobalValue(ServerMeter.NETTY_CONNECTION_RESPONSES_SENT, 1);
-            _serverMetrics.addMeteredGlobalValue(ServerMeter.NETTY_CONNECTION_BYTES_SENT, responseBytes.length);
-            _serverMetrics.addTimedValue(ServerTimer.NETTY_CONNECTION_SEND_RESPONSE_LATENCY, sendResponseLatencyMs,
-                TimeUnit.MILLISECONDS);
-
-            int totalQueryTimeMs = (int) (sendResponseEndTimeMs - queryArrivalTimeMs);
-            if (totalQueryTimeMs > SLOW_QUERY_LATENCY_THRESHOLD_MS) {
-              LOGGER.info(
-                  "Slow query: request handler processing time: {}, send response latency: {}, total time to handle request: {}",
-                  queryProcessingTimeMs, sendResponseLatencyMs, totalQueryTimeMs);
-            }
-          });
+          // responseBytes contains either query results or exception.
+          sendResponse(ctx, queryArrivalTimeMs, responseBytes);
+        } else {
+          // Send exception response.
+          sendResponse(ctx, queryRequest.getRequestId(), queryArrivalTimeMs, new DataTableImplV2(),
+              new Exception("Null query response."));
         }
       }
 
       @Override
       public void onFailure(Throwable t) {
-        LOGGER.error("Caught exception while processing instance request", t);
-        _serverMetrics.addMeteredGlobalValue(ServerMeter.UNCAUGHT_EXCEPTIONS, 1);
+        // Send exception response.
+        LOGGER.error("Exception while processing instance request", t);
+        sendResponse(ctx, instanceRequest.getRequestId(), queryArrivalTimeMs, new DataTableImplV2(), new Exception(t));
       }
     }, MoreExecutors.directExecutor());
   }
 
   @Override
   public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-    LOGGER.error("Caught exception while fetching instance request", cause);
-    _serverMetrics.addMeteredGlobalValue(ServerMeter.REQUEST_FETCH_EXCEPTIONS, 1);
+    // Since we do not know the requestId of the original request here, there is no way for Broker to know which query
+    // request this response belongs to. Hence, Broker will continue to wait for the original request until time out.
+    // To prevent broker from waiting unncessarily,try to catch and handle all exceptions in channelRead0 method so that
+    // this function is never called.
+    LOGGER.error("Unhandled Exception in " + getClass().getCanonicalName(), cause);
+    sendResponse(ctx, 0, System.currentTimeMillis(), new DataTableImplV2(), new Exception(cause));
+  }
+
+  /**
+   * Send an exception back to broker as response to the query request.
+   */
+  private void sendResponse(ChannelHandlerContext ctx, long requestId, long queryArrivalTimeMs, DataTable dataTable,
```

Comment:
Not sure if I understand why we need to separate signatures for this? Can the caller not prepare the DaaTable as necessary for the response, (eg set the exception in the data table if there's one)? And this method is just responsible for sending back the response?

Reply:
The `onSuccess` method in the callback gets byte data so to send that byte data back to server we need a `sendResponse` function that takes byte data as input. Then there are cases where we detect error conditions and need to create our own byte representation of the error condition. This is where the first `sendResponse` method is called. The first `sendResponse` method converts DataTable into byte data and calls the second more generic `sendResponse` method. I can rename the first `sendResponse` method to `sendErrorResponse` if that clarifies things better.

Reply:
I would probably refactor this as two separate utility methods:

- Building the response to send. This can have two flavors, one for success, and one for failure
- Another one just for sending the response. And if this turns out to be just function call, then it might even be inline.

More of suggestion at this point, not a blocker for the PR.

Reply:
I renamed the first function to `sendErrorResponse` for clarity. Although further refactoring can organize the code better, but the refactoring seems a bit outside the scope of this issue. For example, currently, the success (and some failure cases) inside QuerySchedular `_queryScheduler.submit(queryRequest)` and we get the response bytes back through a callback.